teuthology/provision/maas.py

```diff
@@ -0,0 +1,430 @@
+import io
+import json
+import logging
+
+from oauthlib.oauth1 import SIGNATURE_PLAINTEXT
+from requests import Response
+from requests_oauthlib import OAuth1Session
+from typing import Any, Dict, List, Optional, Union, Tuple
+
+import teuthology.orchestra
+
+from teuthology.config import config
+from teuthology.contextutil import safe_while
+from teuthology.orchestra.opsys import OS
+from teuthology import misc
+
+log = logging.getLogger(__name__)
+
+
+def enabled(warn: bool = False) -> bool:
+    """Check for required MAAS settings
+
+    :param warn: Whether to log a message containing unset parameters
+
+    :returns: True if all required settings are present; False otherwise
+    """
+    maas_conf = config.get("maas", {})
+    params: List[str] = ["api_url", "api_key", "machine_types"]
+    unset = [param for param in params if not maas_conf.get(param)]
+    
+    if unset and warn:
+        unset = " ".join(unset)
+        log.warning(
+            f"MAAS disabled; set the following config options to enable: {unset}",
+        )
+
+    if unset:
+        if not config.get("maas", {}).get("api_url"):
+            return False
+
+        api_key = config.get("maas", {}).get("api_key")
+        if not api_key:
+            return False
+        if len(api_key.split(":")) < 3:
+            log.warning(
+                "MAAS api_key appears to be malformed; expected format is "
+                "'consumer_key:consumer_token:secret'"
+            )
+            return False
+
+    return True
+
+def get_types() -> List[str]:
+    """Fetch and parse MAAS machine_types config.
+
+    :returns: The list of MAAS-configured machine types.
+                Returns an empty list if MAAS is not configured
+    """
+    if not enabled():
+        return []
+    maas_conf = config.get("maas", {})
+    types = maas_conf.get("machine_types", "")
+    if not isinstance(types, list):
+        types = types.split(",")
+    return [type_ for type_ in types if type_]
+
+
+def get_session() -> OAuth1Session:
+    """Create an OAuth1Session for communicating with the MAAS server
+    
+    :returns: OAuth1Session: An authenticated session object configured with the
+                MAAS API key credentials
+    """
+    if not enabled():
+        raise RuntimeError("MAAS is not configured!")
+
+    key, token, secret = config.maas["api_key"].split(":")
+    return OAuth1Session(
+        key,
+        resource_owner_key=token,
+        resource_owner_secret=secret,
+        signature_method=SIGNATURE_PLAINTEXT
+    )
+
+
+class MAAS(object):
+    """Reimage machines with https://maas.io"""
+
+    def __init__(
+            self, name: str, os_type: Optional[str] = None, os_version: Optional[str] = None
```

Comment:
Are os_type/os_version really optional? Is that a call to "reprovision how you already are"?

Reply:
Now set default to Ubuntu 22.04.